### PR TITLE
ACRN:DM: Add the pci msi per-vector masking capable

### PR DESCRIPTION
**Pull request by liudlong**
Add msi per-vector masking capable and emulate msi mask
pending mechanism in DM.

Tracked-On: #7265
Signed-off-by: Liu Long <long.liu@linux.intel.com>
Acked-by: Wang, Yu1 <yu1.wang@intel.com>

### CODE REVIEW

**Comment by acrnsi** [comment]
Can one of the admins verify this patch?

File: devicemodel/hw/pci/core.c [hunk 1]
```diff
@@ -1059,6 +1059,79 @@ pci_emul_deinit(struct vmctx *ctx, struct pci_vdev_ops *ops, int bus, int slot,
...
+
+	rc = pci_emul_find_capability(dev, PCIY_MSI, &offset);
+
+	if (msi_cap + offset > dev->prevcap) {
+		pr_err("%s: Msi capability length is out of msi length!\n", __func__);
+		return -1;
+	}
+	if (rc)
+		return -1;
+
```
Comment (ywan170) [inline devicemodel/hw/pci/core.c:1076]: Keep it simple.
if (msi_cap > PCIR_MSI_PENDING) {
    pr_err("%s: invalid offset of msi capability.\n", __func__);
    return -1;
}

File: devicemodel/hw/pci/core.c [hunk 2]
```diff
@@ -1059,6 +1059,79 @@ pci_emul_deinit(struct vmctx *ctx, struct pci_vdev_ops *ops, int bus, int slot,
...
+	int rc;
+
+	rc = pci_access_msi(dev, PCIR_MSI_PENDING, &val, false);
+	if (rc)
+		return 0;
+	return val & (1 << index);
+}
+
+static void
+set_msi_pending_status(struct pci_vdev *dev, uint32_t index, bool set)
```
Comment (ywan170) [inline devicemodel/hw/pci/core.c:1116]: pci_set_msi_pending

**Review by ywan170 — CHANGES_REQUESTED** [review]

File: devicemodel/hw/pci/core.c (hunk 2 above)
Comment (liudlong) [reply devicemodel/hw/pci/core.c]: Update the PR.

File: devicemodel/hw/pci/core.c (hunk 1 above)
Comment (liudlong) [reply devicemodel/hw/pci/core.c:1076]: Update the PR.

**Review by ywan170 — APPROVED** [review]

**Pull request merged**